### ISSUE TITLE
tRNA charging/FBA floating point precision

### ISSUE THREAD

**Issue by thalassemia** [CovertLab/wcEcoli#1427]
It turns out there are some quirks in Numba's handling of `np.sum`, which is used by the `dcdt` function extracted in gh-1425. Take the following `test.py`:
```python
from numba import njit
import numpy as np

c_init = np.array([0.1] * 10) + np.random.default_rng(100).random(10)

@njit
def f_jit(c):
	return np.sum(c)

def f(c):
	return np.sum(c)

print(f_jit(c_init) - f(c_init))
```
Enabling JIT compilation causes the outputs to diverge on my machine:
```
NUMBA_DISABLE_JIT=1 python test.py
>>> 0.0
NUMBA_DISABLE_JIT=0 python test.py
>>> 8.881784197001252e-16
```
Interestingly, the values being summed matter, and replacing the random seed with 0 causes the difference to disappear for me. Numba may just be using a slightly different algorithm for this operation, as noted in [their documentation](https://numba.readthedocs.io/en/stable/reference/fpsemantics.html).

Since our use of `np.sum` occurs in a derivative function passed to `solve_ivp`, this error accumulates until reaching relative errors on the order of 1e-10, very small but enough to dramatically change the solution fluxes calculated by FBA. 

More confusingly, I found that even before gh-1425, there already existed a small amount of run-to-run variation in the tRNA charging and FBA listeners. I'm still searching for the source of that variability, but one thing that might help is to have polypeptide elongation pass integer molecule counts to FBA instead of floating point concentrations.

**Thankfully, these differences do not propagate to the bulk molecule counts or anything else AFAIK, so it should not affect anyone not working on metabolism/FBA.**

**Comment by 1fish2** [comment]
My Intel Mac gets the same results. Testing more variations:

```python
def sum_up(c):
     r = 0.0
     for x in c:
        r += x
     return r

@njit
def sum_up_jit(c):
     r = 0.0
     for x in c:
        r += x
     return r

def sum_up_reverse(c):
     r = 0.0
     for x in c[::-1]:
         r += x
     return r

np.sum(c_init), c_init.sum(), f_jit(c_init), sum(c_init), sum_up(c_init), sum_up_jit(c_init), sum_up_reverse(c_init)
(6.912225035719857,
 6.912225035719857,
 6.912225035719858,
 6.912225035719858,
 6.912225035719858,
 6.912225035719858,
 6.912225035719857)
```

Numpy seems to round differently than a regular Python `sum()` or the `sum_up()` loop (JIT compiled or not). That led to a hypothesis that Numpy is scanning the array in a different order, thus `sum_up_reverse()`. Summing in reverse order (JIT compiled or not) gets the same result as Numpy. **But** repeating the test with random seed `200`, Numpy matched `sum()` and `sum_up()` while `sum_up_reverse()` didn't match (JIT compiled or not).

**Comment by thalassemia** [comment]
Closing with the following summary:

- The run-to-run floating point variance from before this PR disappeared when I upgraded my M2 Macbook Air from macOS 14.2.1 to 14.3. Additionally, I could not reproduce this on my x86 Linux desktop, so I'm not pursuing this any further.
- Numba JIT compilation introduced floating point errors that actually are reproducible across all tested platforms and occasionally result in dramatically different solution fluxes for metabolism, with one flux value jumping from 0 to ~91 mM/s in the optimized model. For scale, the maximum flux observed was ~1300 mM/s, and the mean flux across all reactions and timesteps was ~0.01 mM/s. 
- Despite these flux differences, multiplying the pre- and post-optimization fluxes by the stoichiometric matrix almost always resulted in the same metabolite deltas after rounding. The largest difference I saw before rounding was about 0.13 water molecules. 
- **Since the model uses stochastic rounding, the model with my optimizations can deviate from prior behavior by one molecule here and there, which I am not concerned about.**
- I found no large differences in the kinetic objective values before and after my optimizations, something I would have expected to see if the optimized model was doing a worse job of obeying the kinetic constraints. This leads me to believe that the flux differences are a result of our problem being underconstrained (many different possible solutions with near identical objective values and inputs).
